### PR TITLE
Add EvaluateEntitlement filter

### PR DESCRIPTION
**Pull request by nikosev**

### CODE REVIEW

File: lib/Auth/Process/ΕvaluateΕntitlement.php [hunk 1]
```diff
@@ -0,0 +1,175 @@
+<?php
+
+/**
+ * Attribute filter to evaluate entitlement for access to Services.
+ *
+ */
+class sspmod_entitlement_Auth_Process_ΕvaluateΕntitlement extends SimpleSAML_Auth_ProcessingFilter
```
Comment (NicolasLiampotis) [inline lib/Auth/Process/ΕvaluateΕntitlement.php]: Rename `EvaluateEntitlement` to `AddCapability`

File: lib/Auth/Process/ΕvaluateΕntitlement.php [hunk 2]
```diff
@@ -0,0 +1,175 @@
+<?php
+
+/**
+ * Attribute filter to evaluate entitlement for access to Services.
```
Comment (NicolasLiampotis) [inline lib/Auth/Process/ΕvaluateΕntitlement.php]: "Attribute filter for evaluating the resources capabilities of the authenticating user. 
A capability defines the resource or child-resource the user is allowed to access, optionally specifying certain actions the user is entitled to perform. Capabilities can be used to convey – in a compact form – authorisation information. Capability values are formatted as URNs following the syntax specified in in https://aarc-community.org/guidelines/aarc-g027"

File: lib/Auth/Process/ΕvaluateΕntitlement.php [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+ *
+ */
+class sspmod_entitlement_Auth_Process_ΕvaluateΕntitlement extends SimpleSAML_Auth_ProcessingFilter
+{
+
+    /**
+     * The attribute that will hold the entitlement value.
```
Comment (NicolasLiampotis) [inline lib/Auth/Process/ΕvaluateΕntitlement.php]: The attribute that will hold the capability value(s)

File: lib/Auth/Process/ΕvaluateΕntitlement.php [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+     * @var string
+     */
+    private $attributeName = 'eduPersonEntitlement';
+
+
+    /**
+     * The assigned entitlement value.
```
Comment (NicolasLiampotis) [inline lib/Auth/Process/ΕvaluateΕntitlement.php]: The assigned capability value(s)

File: lib/Auth/Process/ΕvaluateΕntitlement.php [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+     * @var string
+     */
+    private $entitlement = array();
```
Comment (NicolasLiampotis) [inline lib/Auth/Process/ΕvaluateΕntitlement.php]: `entitlement` --> `capability`

File: lib/Auth/Process/ΕvaluateΕntitlement.php [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+
+
+    /**
+     * List of IdP entity IDs excluded from this entitlement.
```
Comment (NicolasLiampotis) [inline lib/Auth/Process/ΕvaluateΕntitlement.php]: List of IdP entity IDs excluded from this capability

File: lib/Auth/Process/ΕvaluateΕntitlement.php [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+     */
+    private $idpBlacklist = array();
+
+
+    /**
+     * List of IdP entity IDs qualifying for this entitlement.
```
Comment (NicolasLiampotis) [inline lib/Auth/Process/ΕvaluateΕntitlement.php]: List of IdP entity IDs qualifying for this capability

File: lib/Auth/Process/ΕvaluateΕntitlement.php [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+     */
+    private $idpWhitelist = array();
+
+
+    /**
+     * Combination of entity attributes qualifying for this entitlement.
```
Comment (NicolasLiampotis) [inline lib/Auth/Process/ΕvaluateΕntitlement.php]: ... qualifying for this entitlement --> qualifying for this capability

File: lib/Auth/Process/ΕvaluateΕntitlement.php [hunk 9]
```diff
@@ -0,0 +1,175 @@
...
+     */
+    private $entityAttributeWhitelist = array();
+
+
+    /**
+     * List of user entitlements qualifying for this entitlement.
```
Comment (NicolasLiampotis) [inline lib/Auth/Process/ΕvaluateΕntitlement.php]: List of user entitlements qualifying for this capability

File: lib/Auth/Process/ΕvaluateΕntitlement.php [hunk 10]
```diff
@@ -0,0 +1,175 @@
...
+     * @throws SimpleSAML_Error_Exception if the mandatory 'accepted' configuration option is missing.
+     */
+    public function __construct($config, $reserved)
+    {
+        parent::__construct($config, $reserved);
+        assert('is_array($config)');
+
+        if (array_key_exists('attributeName', $config)) {
+            if (!is_string($config['attributeName'])) {
+                SimpleSAML_Logger::error("[ΕvaluateΕntitlement] Configuration error: 'attributeName' not a string literal");
```
Comment (NicolasLiampotis) [inline lib/Auth/Process/ΕvaluateΕntitlement.php]: [AddCapability]

File: lib/Auth/Process/ΕvaluateΕntitlement.php [hunk 11]
```diff
@@ -0,0 +1,175 @@
...
+                throw new Exception("ΕvaluateΕntitlement configuration error: 'attributeName' not a string literal");
+            }
+            $this->attributeName = $config['attributeName'];
+        }
+
+        if (array_key_exists('entitlement', $config)) {
```
Comment (NicolasLiampotis) [inline lib/Auth/Process/ΕvaluateΕntitlement.php]: entitlement --> capability

File: lib/Auth/Process/ΕvaluateΕntitlement.php [hunk 12]
```diff
@@ -0,0 +1,175 @@
...
+    {
+        assert('is_array($state)');
+
+        if (!$this->isQualified($state)) {
+            return;
+        }
+        if (empty($state['Attributes'][$this->attributeName])) {
+            $state['Attributes'][$this->attributeName] = array();
+        }
+        $state['Attributes'][$this->attributeName] = array_merge($state['Attributes'][$this->attributeName], $this->entitlement);
```
Comment (NicolasLiampotis) [inline lib/Auth/Process/ΕvaluateΕntitlement.php]: `$this->entitlement` --> `$this-->capability`

Comment (NicolasLiampotis) [reply lib/Auth/Process/ΕvaluateΕntitlement.php]: Add debug: `"[AddCapability] Adding capability " . var_export($this-->capability, true)"`

**Review by NicolasLiampotis — CHANGES_REQUESTED** [review]
- Rename filter to AddCapability
- Add debug log

**Comment by nikosev** [comment]
Change requests have been resolved

**Review by NicolasLiampotis — APPROVED** [review]
LGTM

**Pull request merged**